### ISSUE TITLE
Error generating an ipa: extension without "iPhone-ARM" platform + "hideAneLibSymbols yes"

### ISSUE THREAD

**Issue by caraiz** [airsdk/Adobe-Runtime-Support#1680]
## Problem Description

Generate an ipa with an extension (without iPhone-ARM platform) causes this error:

**Compilation failed while executing : ld64**

It should use the "default" platform without error.

## Steps to Reproduce

The error only happens using the option "-hideAneLibSymbols yes"
With "-hideAneLibSymbols no", the ipa is compiled successfully.

## extension.xml

```
<extension xmlns="http://ns.adobe.com/air/extension/32.0">
	<id>xxx</id>
	<versionNumber>1.0.0</versionNumber>
	<platforms>
		<platform name="Android-ARM">
			<applicationDeployment>...</applicationDeployment>
		</platform>
		<platform name="Android-x86">
			<applicationDeployment>...</applicationDeployment>
		</platform>
		<platform name="default">
			<applicationDeployment/>
		</platform>
	</platforms>
</extension>
```

**Comment by caraiz** [comment]
I use it to avoid symbol conflict using mutiple extensions with iPhone-ARM.

For example, if you have 2 "almost" identical code extensions but with diferent id:

```
FREObject execute(FREContext ctx,void* functionData,uint32_t argc,FREObject argv[]) {return FREString(@"ios-test1-execute");}
FREObject execute1(FREContext ctx,void* functionData,uint32_t argc,FREObject argv[]) {return FREString(@"ios-test1-execute1");}

void ContextInitializer(...) {...}
void ContextFinalizer(...) {...}
void ExtensionInitializer(...) {...}
void ExtensionFinalizer(...) {...}
````

```
FREObject execute(FREContext ctx,void* functionData,uint32_t argc,FREObject argv[]) {return FREString(@"ios-test2-execute");}
FREObject execute2(FREContext ctx,void* functionData,uint32_t argc,FREObject argv[]) {return FREString(@"ios-test2-execute2");}

void ContextInitializer(...) {...}
void ContextFinalizer(...) {...}
void ExtensionInitializer(...) {...}
void ExtensionFinalizer(...) {...}
````

With "-hideAneLibSymbols yes" it works:

Calling "extension1 -> execute" it returns "ios-test1-execute" (ok)
Calling "extension2 -> execute" it returns "ios-test2-execute" (ok)
Calling "extension1 -> execute1" it returns "ios-test1-execute1" (ok)
Calling "extension2 -> execute2" it returns "ios-test2-execute2" (ok)

With "-hideAneLibSymbols no" it generates "Argument Error #3500" (conflict loading the extensions?)


Now, modifying ContextXXX and ExtensionXXX to avoid same name, and "-hideAneLibSymbols no", the #3500 dont happens, but it dont work as expected:

Calling "extension1 -> execute" it returns "ios-test1-execute1" (ok)
Calling "extension2 -> execute" it returns "ios-test1-execute1" (fail, i suppose it uses the first loaded function)
Calling "extension1 -> execute1" it returns "ios-test1-execute1" (ok)
Calling "extension2 -> execute2" it returns "ios-test2-execute2" (ok)


Is it possible avoid names conflict without using "-hideAneLibSymbols yes"?

**Comment by marchbold** [comment]
Oh I totally understand the intent, but really this is an issue with bad function naming by the developer and it means other extensions can't access shared code so you have to replicate functionality across multiple extensions.

Eg our extensions have a shared utility library in one of our extensions that nearly all our other extensions use, and with this option it blocks this approach increasing library size and actually making some functionality impossible.

It was fine in the early days when extensions where isolated native calls, but with the complexity that they have grown I really encourage people not to use this option.

Cheers

**Comment by caraiz** [comment]
I see, its complicated, cause is impossible to know how other people make extensions,
and if only a method is named equals, then i think it will not work as expected...

Imagine an extension which play an audio and other play a video, the logical thing is to have a function called "play" in both.
Then, the "good practice" is have function names as **extension1_play()** and **extension2_play()**?

To avoid possible conflicts with other extensions, you use a preffix in each method?

Thanks

**Comment by marchbold** [comment]
Yes exactly, we prefix every exposed c function with something unique to the extension. 

Anyone who is writing a library that is to be shared with c code, needs to ensure publicly exposed functions are unique. Similar to what you do with packages in actionscript / java when creating a swc / jar. In iOS objective c it is convention to use a prefix on all your classes eg `NSString` (The `NS` is a prefix indicating it's from the foundation library). These same rules do apply to Android and other platforms too, it's just that there they have a system in place for you, i.e. packages, but you do still have to ensure your package names and classes are unique.

In native extensions these c functions are never exposed to the user, and actually only really referenced in your native code and in the initialiser etc in your extension descriptor xml. So you don't even have to worry about usage when naming these functions.

To create a function like "play()" would be terrible as that's way too generic and bound to conflict, if everyone followed such a simple convention. We have approached it with something like: `distriqt_extensionname_functionname` (and even in some cases `distriqt_extension_component_function`).
We use the company name at the beginning to distinguish between similarly named extensions (eg there have been a few called "mediaplayer" and "share" over the years). 


This option really only came about because of lazy developers and Adobe dropped a quick fix in, but it breaks so many other things when it's used that we really wish it was removed :)
